ui/components/ui/numeric-input/numeric-input.component.js

```diff
@@ -26,7 +28,10 @@ export default function NumericInput({
           }
         }}
         onChange={(e) => {
-          onChange?.(parseFloat(e.target.value || 0, 10));
+          const newValue = e.target.value;
+          const match = DECIMAL_REGEX.exec(newValue);
+          if (match[1]?.length >= 15) return;
+          onChange?.(parseFloat(newValue || 0, 10));
```

Comment:
I added above code to ensure that user is not able to enter an invalid numeric value into number input, invalid in sense that it has > 15 decimal places.

---------------------------------------------------------------------------------------------------------------
Number inputs are quite complicated in metamask's scenario. As decimal places in number value exceed 15 the inputs is no longer controlled input in react, it starts behaving in-controlled. Ref: https://github.com/facebook/react/issues/1549

Additional complexity is that `parseFloat` ignores digits after 15 significant digits

parseFloat(1.11111111111111)                          -> 1.111111111111111
parseFloat(1.111111111111111111111111111)    -> 1.111111111111111

Thus user may see that they have added any number of decimal digits but in code it can not exceed 15 - it only looks like value has > 15 decimal places as input becomes un-controlled.

---------------------------------------------------------------------------------------------------------------
Approach I took here is as soon as user enters 15th decimal digit - we reject the change and do not let user enter any more decimal digits. This is to ensure that input does not becomes un-controlled. I expect 14 decimal places to suffice for all use cases.

Adding form error for > 15 decimal places to inputs are also complicated as code outside `onChange` method will see this value `parseFloat(newValue || 0, 10)` which will never exceed 15 decimal places.

---------------------------------------------------------------------------------------------------------------
To allow users to enter more decimal places we need to change the inputs to text inputs.

Reply:
I read through the GH issue you linked to. I am curious whether the input becomes uncontrolled still as was described there as a fix for this landed in React a while back. So, I wonder if the weird behavior we are seeing right now is a combination of a floating point issue (floating numbers only support a certain number of significant digits) and the fact that we are using `parseFloat` here but BigNumber later. 

Regardless, it seems that this is a good change. I am wondering whether we could go farther and have the regex be this:

``` javascript
const DECIMAL_REGEX = /^\d+|\d*\.\d{15,}/u;
```

The reason is that even with this change, since we are using `parseFloat` to normalize the value, if you enter something like an `=`, then the input resets to 0. This regex would prevent the user from entering something non-numeric entirely. Then we wouldn't have to run the value through `parseFloat`, as we already know it's in a numeric format. I made a [CodePen](https://codepen.io/mcmire/pen/vYZagRR) that demonstrates this behavior. What do you think?

Reply:
Input does become uncontrolled - it is indeed weird behaviour - at least on chrome number inputs behave weird if value is not valid number. I spent time debugging it and found that input can show any number of decimal places parseFloat returned only 15 and thus app state had 15 at max.

Reply:
Gotcha, interesting. Good to know, thanks for spending time on this.